File: demos/01-eliza/scripts/test_eliza.py
#!/usr/bin/env python3
"""
Test the ELIZA chatbot with the corrected rules.
Verify that the classic conversation examples work correctly.
"""

import sys
import os

# Add parent directory to path so we can import eliza
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from eliza import ELIZA

def test_conversation():
    """Test the classic ELIZA conversation from Weizenbaum's paper."""

    # Initialize ELIZA
    eliza = ELIZA()

    test_cases = [
        {
            'input': "Men are all alike.",
            'expected_keyword': 'alike',
            'expected_pattern': 'In what way',
            'description': 'Should match "alike" keyword and respond with "In what way?"'
        },
        {
            'input': "They're always bugging us about something or other.",
            'expected_keyword': 'always',
            'expected_pattern': 'Can you think of a specific example',
            'description': 'Should match "always" keyword'
        },
        {
            'input': "Well, my boyfriend made me come here.",
            'expected_keyword': 'my',
            'expected_pattern': 'Your boyfriend',
            'description': 'Should match "my" keyword and transform pronouns'
        },
        {
            'input': "He says I'm depressed much of the time.",
            'expected_keyword': 'sad',
            'expected_pattern': 'I am sorry to hear',
            'description': 'Should match depressed (sad synonym) and give sympathetic response'
        },
        {
            'input': "I need some help, that much seems certain.",
            'expected_keyword': 'desire',
            'expected_pattern': 'What would it mean to you if you got',
            'description': 'Should match "need" (desire synonym)'
        },
        {
            'input': "Perhaps I could learn to get along with my mother.",
            'expected_keyword': 'family',
            'expected_pattern': 'Tell me more about your family',
            'description': 'Should match "mother" (family synonym)'
        },
        {
            'input': "I am sad.",
            'expected_keyword': 'sad',
            'expected_pattern': 'sorry to hear',
            'description': 'Should recognize sad emotion'
        },
        {
            'input': "I want to be happy.",
            'expected_keyword': 'desire',
            'expected_pattern': None,  # Various responses possible
            'description': 'Should match "want" (desire synonym)'
        },
        {
            'input': "I'm sorry.",
            'expected_keyword': 'sorry',
            'expected_pattern': 'apologise',
            'description': 'Should match "sorry" keyword but not too often'
        },
        {
            'input': "Can you help me?",
            'expected_keyword': 'can',
            'expected_pattern': None,
            'description': 'Should match "can" keyword'
        },
        {
            'input': "Why do you ask?",
            'expected_keyword': 'why',
            'expected_pattern': None,
            'description': 'Should match "why" keyword'
        },
        {
            'input': "Everyone hates me.",
            'expected_keyword': 'everyone',
            'expected_pattern': 'Really',
            'description': 'Should match "everyone" keyword'
        },
        {
            'input': "You are like my father.",
            'expected_keyword': 'like',
            'expected_pattern': None,
            'description': 'Should match "like" keyword with @be pattern'
        },
        {
            'input': "I remember my childhood.",
            'expected_keyword': 'remember',
            'expected_pattern': 'Do you often think',
            'description': 'Should match "remember" keyword'
        },
        {
            'input': "If only I could be better.",
            'expected_keyword': 'if',
            'expected_pattern': None,
            'description': 'Should match "if" keyword'
        },
        {
            'input': "I dreamed about you.",
            'expected_keyword': 'dreamed',
            'expected_pattern': None,
            'description': 'Should match "dreamed" keyword'
        },
        {
            'input': "What is your name?",
            'expected_keyword': 'name',
            'expected_pattern': 'not interested in names',
            'description': 'Should match "name" keyword'
        },
        {
            'input': "Do computers think?",
            'expected_keyword': 'computer',
            'expected_pattern': 'computers worry you',
            'description': 'Should match "computer" keyword'
        },
    ]

    print("=" * 80)
    print("TESTING ELIZA WITH CORRECTED RULES")
    print("=" * 80)
    print()

    passed = 0
    failed = 0

    for i, test in enumerate(test_cases, 1):
        print(f"Test {i}: {test['description']}")
        print(f"  Input: {test['input']}")

        response = eliza.respond(test['input'])
        print(f"  Response: {response}")

        # Check if expected pattern appears in response
        if test['expected_pattern']:
            if test['expected_pattern'].lower() in response.lower():
                print(f"  ✓ PASS: Found expected pattern '{test['expected_pattern']}'")
                passed += 1
            else:
                print(f"  ✗ FAIL: Expected pattern '{test['expected_pattern']}' not found")
                failed += 1
        else:
            # Just verify we got a response
            if response and len(response) > 0:
                print(f"  ✓ PASS: Got valid response")
                passed += 1
            else:
                print(f"  ✗ FAIL: No response")
                failed += 1

        print()

    print("=" * 80)
    print(f"RESULTS: {passed} passed, {failed} failed out of {passed + failed} tests")
    print("=" * 80)

    return passed, failed

def test_sorry_priority():
    """Test that 'sorry' doesn't match too frequently."""

    eliza = ELIZA()

    print("\n" + "=" * 80)
    print("TESTING 'SORRY' KEYWORD PRIORITY")
    print("=" * 80)
    print()
    print("Testing that 'sorry' doesn't incorrectly match non-apology statements...")
    print()

    # These should NOT trigger the sorry keyword
    non_apology_inputs = [
        "I am sad and depressed.",
        "I feel unhappy.",
        "I need help.",
        "Can you help me?",
        "I want to be better.",
        "Everyone hates me.",
        "I am lonely.",
        "Why do you ask?",
        "I don't know what to do.",
        "I feel lost.",
    ]

    sorry_count = 0
    for inp in non_apology_inputs:
        response = eliza.respond(inp)
        if 'apologis' in response.lower() or 'apologies' in response.lower():
            print(f"  ✗ FAIL: '{inp}' triggered sorry response: {response}")
            sorry_count += 1
        else:
            print(f"  ✓ PASS: '{inp}' -> {response}")

    if sorry_count == 0:
        print("\n✓ All non-apology inputs avoided 'sorry' response!")
    else:
        print(f"\n✗ {sorry_count} non-apology inputs incorrectly triggered 'sorry' response")

    # These SHOULD trigger the sorry keyword
    print("\nTesting that actual apologies DO trigger 'sorry' response...")
    print()

    apology_inputs = [
        "I'm sorry.",
        "Sorry about that.",
        "I apologize.",
        "I apologise.",
    ]

    for inp in apology_inputs:
        response = eliza.respond(inp)
        if 'apologis' in response.lower() or 'apologies' in response.lower():
            print(f"  ✓ PASS: '{inp}' -> {response}")
        else:
            print(f"  ✗ FAIL: '{inp}' should have triggered sorry response: {response}")

    print()

def interactive_mode():
    """Run ELIZA in interactive mode for manual testing."""

    eliza = ELIZA()

    print("\n" + "=" * 80)
    print("INTERACTIVE MODE")
    print("=" * 80)
    print()
    print(eliza.get_initial_greeting())
    print()
    print("Type 'quit' to exit.")
    print()

    while True:
        try:
            user_input = input("You: ").strip()
            if not user_input:
                continue

            response = eliza.respond(user_input)
            print(f"ELIZA: {response}")
            print()

        except (KeyboardInterrupt, EOFError):
            print("\n\n" + eliza.get_final_greeting())
            break

if __name__ == '__main__':
    # Run automated tests
    passed, failed = test_conversation()
    test_sorry_priority()

    # Optionally run interactive mode
    if len(sys.argv) > 1 and sys.argv[1] == '--interactive':
        interactive_mode()
    else:
        print("\nRun with --interactive flag to test manually.")

    # Exit with error code if tests failed
    sys.exit(0 if failed == 0 else 1)

File: demos/01-eliza/test-eliza.mjs
#!/usr/bin/env node
/**
 * Comprehensive test suite for ELIZA (Demo 01)
 * Tests pattern matching, substitutions, and conversation flow
 * Run with: node test-eliza.mjs
 */

import { readFile } from 'fs/promises';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Mock DOM globals for browser-based code
global.window = {};
global.document = {};

// Load the required modules
const patternMatcherCode = await readFile(join(__dirname, 'js/pattern-matcher.js'), 'utf-8');
const elizaEngineCode = await readFile(join(__dirname, 'js/eliza-engine.js'), 'utf-8');

// Execute PatternMatcher code
global.module = { exports: {} };
eval(patternMatcherCode);
const PatternMatcher = global.module.exports;

if (!PatternMatcher) {
    throw new Error('Failed to load PatternMatcher class');
}

// Make PatternMatcher globally available for ElizaEngine
global.PatternMatcher = PatternMatcher;

// Execute ElizaEngine code
global.module = { exports: {} };
eval(elizaEngineCode);
const ElizaEngine = global.module.exports;

if (!ElizaEngine) {
    throw new Error('Failed to load ElizaEngine class');
}

class TestRunner {
    constructor() {
        this.passed = 0;
        this.failed = 0;
        this.total = 0;
    }

    assert(condition, testName, expected, actual) {
        this.total++;
        if (condition) {
            console.log(`✓ PASS: ${testName}`);
            this.passed++;
        } else {
            console.log(`✗ FAIL: ${testName}`);
            console.log(`  Expected: ${JSON.stringify(expected)}`);
            console.log(`  Actual: ${JSON.stringify(actual)}`);
            this.failed++;
        }
    }

    assertEqual(actual, expected, testName) {
        this.assert(actual === expected, testName, expected, actual);
    }

    assertContains(text, substring, testName) {
        const condition = text.toLowerCase().includes(substring.toLowerCase());
        this.assert(condition, testName, `contains "${substring}"`, text);
    }

    assertNotEmpty(value, testName) {
        this.assert(value && value.length > 0, testName, 'non-empty', value);
    }

    printSummary() {
        console.log('\n' + '='.repeat(70));
        console.log(`Test Summary: ${this.passed}/${this.total} passed, ${this.failed} failed`);
        console.log('='.repeat(70));
        return this.failed === 0;
    }
}

async function runTests() {
    console.log('='.repeat(70));
    console.log('ELIZA Test Suite (Demo 01)');
    console.log('='.repeat(70));
    console.log();

    const runner = new TestRunner();
    const eliza = new ElizaEngine();

    // Load rules
    const rulesData = JSON.parse(await readFile(join(__dirname, 'data/eliza-rules.json'), 'utf-8'));
    await eliza.loadRules(rulesData);

    // Test 1: Pattern Matching - Classic conversation from original paper
    console.log('\n--- Test Group 1: Classic Pattern Matching ---\n');

    const classicTests = [
        {
            input: "Men are all alike.",
            expectedKeyword: "alike",
            description: "Test 'alike' keyword pattern"
        },
        {
            input: "They're always bugging us about something or other.",
            expectedKeyword: "always",
            description: "Test 'always' keyword pattern"
        },
        {
            input: "Well, my boyfriend made me come here.",
            expectedKeyword: "my",
            description: "Test possessive 'my' keyword"
        },
        {
            input: "I need some help, that much seems certain.",
            expectedKeyword: "i",
            description: "Test 'I' pronoun keyword"
        }
    ];

    for (const test of classicTests) {
        const response = eliza.getResponse(test.input);
        runner.assertEqual(
            response.matchInfo.keyword,
            test.expectedKeyword,
            test.description
        );
        runner.assertNotEmpty(response.response, `${test.description} - generates non-empty response`);
    }

    eliza.reset();

    // Test 2: Sorry Keyword Priority
    console.log('\n--- Test Group 2: Keyword Priority (Sorry) ---\n');

    const sorryTests = [
        { input: "I'm sorry.", shouldMatchSorry: true },
        { input: "I apologise.", shouldMatchSorry: true },
        { input: "I am sad and sorry.", shouldMatchSorry: true },
        { input: "I feel unhappy.", shouldMatchSorry: false },
        { input: "I need help.", shouldMatchSorry: false }
    ];

    for (const test of sorryTests) {
        const response = eliza.getResponse(test.input);
        const matchedSorry = response.matchInfo &&
            (response.matchInfo.keyword === 'sorry' || response.matchInfo.keyword === 'apologise');
        runner.assertEqual(
            matchedSorry,
            test.shouldMatchSorry,
            `"${test.input}" ${test.shouldMatchSorry ? 'should' : 'should not'} match sorry/apologise`
        );
    }

    eliza.reset();

    // Test 3: Response Formatting
    console.log('\n--- Test Group 3: Response Formatting ---\n');

    const formatTests = [
        "I feel happy.",
        "Tell me about computers.",
        "Can you help me?"
    ];

    for (const input of formatTests) {
        const response = eliza.getResponse(input);
        const isAllUppercase = response.response === response.response.toUpperCase();
        const isSentenceCase = response.response.charAt(0) === response.response.charAt(0).toUpperCase();

        runner.assert(
            !isAllUppercase || response.matchInfo.keyword === 'fallback',
            `"${input}" - Should not be all uppercase (unless fallback)`,
            'not all uppercase',
            response.response
        );

        runner.assert(
            isSentenceCase,
            `"${input}" - Should start with capital letter`,
            'sentence case',
            response.response
        );
    }

    eliza.reset();

    // Test 4: Pre-substitutions
    console.log('\n--- Test Group 4: Pre-substitutions ---\n');

    const patternMatcher = new PatternMatcher();
    const preSubTests = [
        { input: "don't", expected: "do not" },
        { input: "can't", expected: "can not" },
        { input: "won't", expected: "will not" },
        { input: "I'm", expected: "i am" }
    ];

    for (const test of preSubTests) {
        const result = patternMatcher.applyPreSubstitutions(test.input, rulesData.preSubstitutions);
        runner.assertContains(
            result.result,
            test.expected,
            `Pre-substitution: "${test.input}" → "${test.expected}"`
        );
    }

    // Test 5: Post-substitutions (Reflection)
    console.log('\n--- Test Group 5: Post-substitutions (Reflection) ---\n');

    const postSubTests = [
        { input: "I am", expected: "you are" },
        { input: "my problem", expected: "your problem" },
        { input: "myself", expected: "yourself" },
        { input: "you are", expected: "I am" }
    ];

    for (const test of postSubTests) {
        const result = patternMatcher.applyPostSubstitutions(test.input, rulesData.postSubstitutions);
        runner.assertContains(
            result.result,
            test.expected,
            `Post-substitution: "${test.input}" → "${test.expected}"`
        );
    }

    // Test 6: Wildcard Matching
    console.log('\n--- Test Group 6: Wildcard Pattern Matching ---\n');

    const wildcardTests = [
        {
            pattern: "* I need *",
            input: "I think I need some help",
            shouldMatch: true
        },
        {
            pattern: "* you *",
            input: "Sometimes you make me angry",
            shouldMatch: true
        },
        {
            pattern: "I * you",
            input: "I love you",
            shouldMatch: true
        }
    ];

    for (const test of wildcardTests) {
        const result = patternMatcher.matchPattern(test.input.toLowerCase(), test.pattern, rulesData.synonyms);
        runner.assertEqual(
            result.matched,
            test.shouldMatch,
            `Pattern "${test.pattern}" ${test.shouldMatch ? 'should match' : 'should not match'} "${test.input}"`
        );
    }

    // Test 7: Conversation State
    console.log('\n--- Test Group 7: Conversation State Management ---\n');

    eliza.reset();
    const greeting = eliza.getGreeting();
    runner.assertNotEmpty(greeting, "Initial greeting should not be empty");

    const history1 = eliza.getHistory();
    runner.assertEqual(history1.length, 1, "History should have 1 entry after greeting");

    eliza.getResponse("Hello");
    const history2 = eliza.getHistory();
    runner.assertEqual(history2.length, 3, "History should have 3 entries after one exchange");

    eliza.reset();
    const history3 = eliza.getHistory();
    runner.assertEqual(history3.length, 0, "History should be empty after reset");

    // Test 8: Quit Detection
    console.log('\n--- Test Group 8: Quit Word Detection ---\n');

    eliza.reset();
    const quitResponse = eliza.getResponse("quit");
    runner.assertEqual(quitResponse.isQuit, true, "Should detect 'quit' as quit word");

    eliza.reset();
    const byeResponse = eliza.getResponse("goodbye");
    runner.assertEqual(byeResponse.isQuit, true, "Should detect 'goodbye' as quit word");

    eliza.reset();
    const normalResponse = eliza.getResponse("I am happy");
    runner.assertEqual(normalResponse.isQuit, false, "Normal input should not be quit");

    // Test 9: Rules Validation
    console.log('\n--- Test Group 9: Rules Structure Validation ---\n');

    runner.assert(
        Array.isArray(rulesData.rules),
        "Rules data should contain rules array",
        "array",
        typeof rulesData.rules
    );

    runner.assert(
        rulesData.rules.length > 0,
        "Rules array should not be empty",
        "> 0",
        rulesData.rules.length
    );

    let allRulesValid = true;
    for (const rule of rulesData.rules) {
        if (!rule.keyword || !rule.patterns || !Array.isArray(rule.patterns)) {
            allRulesValid = false;
            break;
        }
        for (const pattern of rule.patterns) {
            if (!pattern.pattern || !pattern.responses || !Array.isArray(pattern.responses)) {
                allRulesValid = false;
                break;
            }
        }
    }
    runner.assert(allRulesValid, "All rules should have valid structure", "valid", allRulesValid ? "valid" : "invalid");

    // Test 10: Synonym Expansion
    console.log('\n--- Test Group 10: Synonym Expansion ---\n');

    // Test that synonyms are properly expanded in patterns
    const synonymTest = eliza.getResponse("I believe in ghosts");
    runner.assertNotEmpty(synonymTest.response, "Should handle 'believe' (synonym of 'think')");

    const familyTest = eliza.getResponse("Tell me about your mother");
    runner.assertNotEmpty(familyTest.response, "Should handle 'mother' (family synonym)");

    // Test 11: Response Cycling
    console.log('\n--- Test Group 11: Response Cycling ---\n');

    eliza.reset();
    const responses = [];
    for (let i = 0; i < 5; i++) {
        const resp = eliza.getResponse("I remember something");
        responses.push(resp.response);
    }

    // Check that we get different responses (cycling through)
    const uniqueResponses = new Set(responses);
    runner.assert(
        uniqueResponses.size > 1 || responses.length === 1,
        "Should cycle through multiple responses or have single response",
        "> 1 unique",
        `${uniqueResponses.size} unique`
    );

    // Final summary
    const success = runner.printSummary();
    process.exit(success ? 0 : 1);
}

// Run the tests
runTests().catch(error => {
    console.error('Error running tests:', error);
    process.exit(1);
});
